### DISCUSSION TITLE
use mongoose as http server, windows as http client ,when i use mg_http_printf_chunk send data 200/s，windows cmd curl occur "Illegal or missing hexadecimal sequence in chunked-encoding", error

### DISCUSSION THREAD

**Discussion by guangnian1126** [cesanta/mongoose#2126] [Ask questions, get answers] [answered]
use mongoose as http server, windows cmd as http client ,when i use mg_http_printf_chunk send data 200/s,contiue about 10 minute，windows cmd curl occur "Illegal or missing hexadecimal sequence in chunked-encoding", error 
code:

int
gmcgi_main(int argc, char **argv)
{
	int ret = 0;
	struct mg_mgr mgr;
	struct mg_connection *nc;

	signal(SIGINT, signal_int);
	signal(SIGQUIT, signal_int);
	signal(SIGTERM, signal_int);

	if (argc == 2 && !strcmp(argv[1], "daemon")) {

		if (daemon(1, 1) < 0) {
			perror("error daemon...\n");
			exit(1);
		}
	}

	if (argc == 3 && !strcmp(argv[1], "daemon")) {
		if (daemon(1, 1) < 0) {
			perror("error daemon...\n");
			exit(1);
		}
		log_set_level(atoi(argv[2]));
	}

	mg_mgr_init(&mgr);
	mg_http_listen(&mgr, s_http_addr, ev_handler, NULL);

	LOGR(INFO, "Starting RESTful server on %s\n", s_http_addr);

	cgi = create_cgi("CGI");
	register_cgi_module(cgi);
	cgi_init_func(cgi);
	while (!signo) {
		mg_mgr_poll(&mgr, 5);
	}

	cgi_destory_func(cgi);
	free_cgi(cgi);

	mg_mgr_free(&mgr);

	LOGR(CRIT, "exit cgi process");

	return 0;
}

static void *thread_steam_post_tag_httpclient(void *param)
{
	struct mg_connection *nc;
	nc = (struct mg_connection *) param;
	while(1) 
	{
		cniot_msg_t* p_msg = cniot_queue_pop_msg(rfid_inventory_tag_queue);
		switch(p_msg->what) 
		{
			case HTTP_MSG_QUEUE_ID: 
			{
				uint8_t* buf = p_msg->obj;
				int read_data_len = p_msg->arg1;
				if (read_data_len >= 0) 
				{
					mg_http_printf_chunk(nc, "%s\r\n", buf);
				}
				break;
			}
			default:
				break;
		}
		cniot_recycle_msg(p_msg);
	}
}

**Comment by scaprile** [comment]
What _exactly_ are you doing ? What is your event handler function ? What are those `cgi_something` functions ? What is that `kniot` stuff ? What is that ´queue´stuff Did you sniff the transfer ? Did you read the docs ? Are you doing things as told ?
http://mongoose.ws/documentation/
Did you see the examples ? Did you read the tutorials ?
Mongoose functions must be called from inside an event handler, all in a single thread. I don't see that. I don't understand anything of what you are doing.

The issue template exists for a reason, **_please honor it and fill it_**.

**Reply by guangnian1126** [reply]
Sorry,Because only a portion of the code was given, it caused you trouble. now  I I followed you said  "Mongoose functions must be called from inside an event handler"。
I want to implement an http server that regularly receives an http request and continuously sends data at a high frequency。
Http server sends 200 messages to clinet per second, with 250 bytes per message.
The core code of the implementation is
```
static void ev_handler(struct mg_connection *c, int ev, void *ev_data, void *fn_data)
{
	size_t len;
	char path[256];
	struct mg_http_message *hm = (struct mg_http_message *)ev_data;
	struct in_addr remote_ip = {0};

	switch (ev) {
		case MG_EV_HTTP_MSG:
               /*Pseudocode*/
                   1 、Parse http requests. If it is legal, call mg_printf function reply， set http connection chunked,as follows
	           //mg_printf(nc, "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nTransfer-Encoding: chunked\r\n\r\n");
                   2 、continuously put the collected data into the message queue
		case MG_EV_POLL:
                /*Pseudocode*/
                    Message queue reads data and calls mg_http_printf_chunk API interface to send。
		default:
			break;
	}
}
        mg_mgr_init(&mgr);
	mg_http_listen(&mgr, s_http_addr, ev_handler, NULL);
	while (!signo) {
		mg_mgr_poll(&mgr, 5);
	}



```
 I have 2 questions.

1 To quickly process message queue data, mg_ mgr_ poll(&mgr, 5);

Event loop is 5ms, is this configuration OK? Is there anything I need to pay attention to?



2 At MG_ EV_ POLL handles reading message queue data. Can I call mg_http_printf_chunk  multiple times frequently ?

**Comment by scaprile** [comment, marked as answer]
Why do you want to send 200 msgs/s ? Why such short messages ? TCP has been designed for anything other than that, and HTTP does not work best like that.
You are putting an unnecessary burden on your system.
There is no need to queue anything, if you have the response, just send it, TCP will control data flow with the other end.
The poll event is not timely regular, it will be triggered every time the manager is woken up, either by a timeout or by data in a socket.

**_Please read the documentation, read the tutorials, and `fill the issue template`_**
